File: saltblock/src/main/java/io/mjolnir/saltblock/SaltBlock.java
package io.mjolnir.saltblock;

import java.security.PublicKey;
import java.util.List;

public class SaltBlock {

    private final EncryptionAlgorithm mAlg;

    public SaltBlock() {
        this(EncryptionAlgorithm.AES);
    }

    /**
     * @param alg specifies which encryption algorithm you want to use. AES is default
     */
    public SaltBlock(EncryptionAlgorithm alg) {
        this.mAlg = alg;
    }

    /**
     * @param plainTexts to encrypt
     * @return return list of cipher texts
     * Note that this uses the default encryption method AES/GCM/NoPadding
     */
    public List<String> encrypt(String keyAlias, List<String> plainTexts) {

        switch (mAlg) {
            case AES:
                try {
                    return AES.encrypt(keyAlias, plainTexts);
                } catch (Exception e) {
                    e.printStackTrace();
                    return null;
                }
            case RSA:
                try {
                    return RSA.encrypt(keyAlias, plainTexts);
                } catch (Exception e) {
                    e.printStackTrace();
                    return null;
                }
            default:
               return  null;
        }
    }

    public String encrypt(String keyAlias, String plainText) {

        switch (mAlg) {
            case AES:
                return AES.encrypt(keyAlias, plainText);
            case RSA:
                return RSA.encrypt(keyAlias, plainText);
            default:
                return  null;
        }
    }

    public List<String> decrypt(String keyAlias, List<String> cipherTexts) {
        switch (mAlg) {
            case AES:
                try {
                    return AES.decrypt(keyAlias, cipherTexts);
                } catch (Exception e) {
                    e.printStackTrace();
                    return null;
                }
            case RSA:
                try {

                    return RSA.decrypt(keyAlias, cipherTexts);
                } catch (Exception e) {
                    e.printStackTrace();
                    return null;
                }
            default:
                return null;
        }
    }

    public String decrypt(String keyAlias, String cipherText) {
        switch (mAlg) {
            case AES:
                return AES.decrypt(keyAlias, cipherText);
            case RSA:
                return RSA.decrypt(keyAlias, cipherText);
            default:
                return null;
        }
    }

    /**
     * @return RSA public key. Returns null if RSA is not implemented
     */
    public PublicKey getPublicKey(String keyAlias) {
       if (mAlg == EncryptionAlgorithm.AES) return null;

       return RSA.getPublicKey(keyAlias);
    }

}

File: saltblock/src/main/java/io/mjolnir/saltblock/SaltBlock.kt
package io.mjolnir.saltblock

import androidx.core.util.Pair
import java.io.File
import java.io.Serializable

class SaltBlock {

    fun encryptAES(keyAlias: String, plainText: String) : String {
        return processEncryptionRequest(keyAlias, plainText)
    }

    fun encryptAES(keyAlias: String, obj: Serializable) : String {
        return processEncryptionRequest(keyAlias, obj)
    }

    fun encryptAES(keyAlias: String, plainTexts: List<String>) : List<String> {
        return processEncryptionRequest(keyAlias, plainTexts)
    }

    fun encryptAES(keyAlias: String, file: File) : File {
        return processEncryptionRequest(keyAlias, file)
    }

    fun encryptRSA(plainText: String, publicKey: String) : String {
        return processRSAEncryptionRequest(plainText, publicKey)
    }

    fun encryptRSA(obj: Serializable, publicKey: String) : String {
        return processRSAEncryptionRequest(obj, publicKey)
    }

    fun encryptRSA(plainTexts: List<String>, publicKey: String) :
            List<String> {
        return processRSAEncryptionRequest(plainTexts, publicKey)
    }

    fun encryptSharedFile(publicKey: String, file: File) :
            Pair<String, File> {
        return processSharedFileEncryptionRequest(publicKey, file)
    }

    fun decryptAES(keyAlias: String, cipherText: String) : String {
        return processDecryptionRequest(EncryptionAlgorithm.AES, keyAlias, cipherText)
    }

    fun decryptToObjAES(keyAlias: String, cipherText: String) : Any {
        return processDecryptionToObjectRequest(EncryptionAlgorithm.AES, keyAlias, cipherText)
    }
    
    fun decryptAES(keyAlias: String, cipherTexts: List<String>) : List<String> {
        return processDecryptionRequest(EncryptionAlgorithm.AES, keyAlias, cipherTexts)
    }

    fun decryptAES(keyAlias: String, file: File) : File {
        return processDecryptionRequest(keyAlias, file)
    }

    fun decryptRSA(keyAlias: String, cipherText: String) : String {
        return processDecryptionRequest(EncryptionAlgorithm.RSA, keyAlias, cipherText)
    }

    fun decryptToObjRSA(keyAlias: String, cipherText: String) : Any {
        return processDecryptionToObjectRequest(EncryptionAlgorithm.RSA, keyAlias, cipherText)
    }

    fun decryptRSA(keyAlias: String, cipherTexts: List<String>) : List<String> {
        return processDecryptionRequest(EncryptionAlgorithm.RSA, keyAlias, cipherTexts)
    }

    fun decryptSharedFile(rsaKeyAlias: String, wrappedKey: String, file: File) : File {
        return processDecryptionRequest(rsaKeyAlias, wrappedKey, file)
    }

    fun getPublicKey(keyAlias: String) : String {
        return processPublicKeyRequest(keyAlias)
    }
}
